### PR TITLE
Precise closure-capture of local variables in the presence of deferred nested functions. This is a good thing to do anyway (improves CQ and solves some odd memory leak cases) and is especially important for redeferral and it allows consistency of byte code regardless of deferral. (Byte code will not necessarily be identical, but it will be functionally the same.) The parser is now responsible for detecting and marking non-local references to user variables, and we no longer force closure-capture of all locals in the presence of a deferred child function.

### PR DESCRIPTION
**Pull request by pleath**
Precise closure-capture of local variables in the presence of deferred
nested functions. This is a good thing to do anyway (improves CQ and
solves some odd memory leak cases) and is especially important for
redeferral and it allows consistency of byte code regardless of
deferral. (Byte code will not necessarily be identical, but it will be
functionally the same.) The parser is now responsible for detecting and
marking non-local references to user variables, and we no longer force
closure-capture of all locals in the presence of a deferred child
function.

### CODE REVIEW

**Comment by pleath** [comment]
@ianwjhalliday @aneeshdk

File: lib/Backend/amd64/LowererMDArch.cpp [hunk 1]
```diff
@@ -414,7 +414,7 @@ LowererMDArch::LoadFuncExpression(IR::Instr *instrFuncExpr)
         paramOpnd = IR::SymOpnd::New(paramSym, TyMachReg, this->m_func);
     }
 
-    if (this->m_func->GetJnFunction()->IsGenerator())
+    if (instrFuncExpr->m_func->GetJnFunction()->IsGenerator())
```
Comment (ianwjhalliday) [inline lib/Backend/amd64/LowererMDArch.cpp]: Is this a drive by bug fix?

Comment (pleath) [inline lib/Backend/amd64/LowererMDArch.cpp]: Yes and no. It’s not directly related to the feature, but there are a few things that fix bugs the feature exposed in the UT’s, mainly by enabling more optimization.

--Paul

File: lib/Parser/Parse.cpp [hunk 2]
```diff
@@ -1701,6 +1703,12 @@ void Parser::BindPidRefsInScope(IdentPtr pid, Symbol *sym, int blockId, uint max
             sym->SetIsModuleExportStorage(true);
         }
 
+        if (ref->GetFuncScopeId() != funcId && !sym->GetIsGlobal())
+        {
+            Assert(ref->GetFuncScopeId() > funcId);
```
Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: So this means that the ref here is coming from a nested function, correct?

I'm not following what this is always the case, versus a ref coming from an enclosing scope.  I will admit my understanding of the PidRef model is not good.

File: lib/Parser/Parse.cpp [hunk 3]
```diff
@@ -4505,48 +4486,21 @@ ParseNodePtr Parser::ParseFncDecl(ushort flags, LPCOLESTR pNameHint, const bool
...
-        m_currentNodeDeferredFunc = pnodeFncSave;
-        if (m_currentNodeFunc && pnodeFnc->sxFnc.HasWithStmt())
-        {
-            GetCurrentFunctionNode()->sxFnc.SetHasWithStmt(true);
-        }
-    }
+    // Lambdas do not have "arguments" and instead capture their parent's
+    // binding of "arguments.  To ensure the arguments object of the enclosing
+    // non-lambda function is loaded propagate the UsesArguments flag up to
+    // the parent function
```
Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: Why did this comment move up here?  It doesn't mention eval at all, and this code here isn't specific to lambdas

File: lib/Parser/Parse.cpp [hunk 4]
```diff
@@ -4713,15 +4671,72 @@ bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, ParseNodePtr pnodeFncPare
...
+        {
+            AnalysisAssert(pnodeFnc);
+
+            if (!fLambda)
+            {
+                m_currentNodeNonLambdaDeferredFunc = pnodeFnc;
+            }
+
+            m_currentNodeDeferredFunc = pnodeFnc;
+        }
```
Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: This code looks the same as above, could be refactored into a helper function

File: lib/Parser/Parse.cpp [hunk 5]
```diff
@@ -5243,6 +5258,36 @@ bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, ParseNodePtr pnodeFncPare
...
+        if (!fLambda)
+        {
+            Assert(pnodeFnc == m_currentNodeNonLambdaDeferredFunc);
+            m_currentNodeNonLambdaDeferredFunc = pnodeFncSaveNonLambda;
+        }
+        m_currentNodeDeferredFunc = pnodeFncSave;
+        if (m_currentNodeFunc && pnodeFnc->sxFnc.HasWithStmt())
+        {
+            GetCurrentFunctionNode()->sxFnc.SetHasWithStmt(true);
+        }
```
Comment (ianwjhalliday) [inline lib/Parser/Parse.cpp]: Huh, why is this only propagated in the deferred parse case?

Also seems out of place here, as opposed to to where we propagate callseval and hasarguments flags above, at the call site to ParseFncDeclHelper

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 6]
```diff
@@ -2536,7 +2541,7 @@ void ByteCodeGenerator::GetEnclosingNonLambdaScope(FuncInfo *funcInfo, Scope * &
         {
             envIndex++;
         }
-        if (((scope == scope->GetFunc()->GetBodyScope() || scope == scope->GetFunc()->GetParamScope()) && !scope->GetFunc()->IsLambda()) || scope->IsGlobalEvalBlockScope())
+        if ((((scope == scope->GetFunc()->GetBodyScope() && (!scope->GetFunc()->GetParamScope() || scope->GetFunc()->GetParamScope()->GetCanMergeWithBodyScope())) || scope == scope->GetFunc()->GetParamScope()) && !scope->GetFunc()->IsLambda()) || scope->IsGlobalEvalBlockScope())
```
Comment (ianwjhalliday) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: Is this an unrelated bug fix or was this broken by the precise closure-capture somehow?

File: lib/Runtime/ByteCode/ByteCodeGenerator.cpp [hunk 7]
```diff
@@ -2695,6 +2691,12 @@ FuncInfo* PostVisitFunction(ParseNode* pnode, ByteCodeGenerator* byteCodeGenerat
         return top;
     }
 
+    if (top->paramScope && top->paramScope->GetCanMergeWithBodyScope())
+    {
+        Scope::MergeParamAndBodyScopes(pnode);
+        Scope::RemoveParamScope(pnode);
+    }
```
Comment (ianwjhalliday) [inline lib/Runtime/ByteCode/ByteCodeGenerator.cpp]: Is this a change to how we handle the param scope that was necessary for precise capture?

I'm concerned that this may have bugs and if it is not specific to precise capture, would be better to have in a separate change.

File: lib/Backend/amd64/LowererMDArch.cpp (hunk 1 above)
Comment (ianwjhalliday) [inline lib/Backend/amd64/LowererMDArch.cpp]: Gotcha

**Comment by ianwjhalliday** [comment]
Awesome to see our closure capture becoming precise.

Mostly I had questions rather than comments. Overall looks good to me.

**Comment by ianwjhalliday** [comment]
One of @Cellule @MikeHolman should take a quick look at the asmjs change.

**Comment by MikeHolman** [comment]
asm.js change lgtm

File: lib/Parser/Parse.cpp [hunk 8]
```diff
@@ -1679,6 +1680,7 @@ void Parser::BindPidRefs(BlockInfoStack *blockInfo, uint maxBlockId)
 void Parser::BindPidRefsInScope(IdentPtr pid, Symbol *sym, int blockId, uint maxBlockId)
 {
     PidRefStack *ref, *nextRef, *lastRef = nullptr;
+    Js::LocalFunctionId funcId = GetCurrentFunctionNode()->sxFnc.functionId;
     Assert(sym);
 
     for (ref = pid->GetTopRef(); ref && ref->GetScopeId() >= blockId; ref = nextRef)
```
Comment (aneeshdk) [inline lib/Parser/Parse.cpp]: Not related to change. Do we need both >= condition here and the equality check at the end of the loop body to break out of the loop?

File: lib/Parser/Parse.cpp [hunk 9]
```diff
@@ -4693,8 +4647,12 @@ void Parser::AppendFunctionToScopeList(bool fDeclaration, ParseNodePtr pnodeFnc)
 Parse a function definition.
 ***************************************************************************/
 template<bool buildAST>
-bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, ParseNodePtr pnodeFncParent, LPCOLESTR pNameHint, ushort flags, bool *pHasName, bool fUnaryOrParen, bool noStmtContext, bool *pNeedScanRCurly)
+bool Parser::ParseFncDeclHelper(ParseNodePtr pnodeFnc, LPCOLESTR pNameHint, ushort flags, bool *pHasName, bool fUnaryOrParen, bool noStmtContext, bool *pNeedScanRCurly)
 {
+    ParseNodePtr pnodeFncSave = buildAST ? m_currentNodeFunc : m_currentNodeDeferredFunc;
```
Comment (aneeshdk) [inline lib/Parser/Parse.cpp]: Will calling GetCurrentFunctionNode() work?

Comment (aneeshdk) [inline lib/Parser/Parse.cpp]: Similarly GetCurrentNonLamdaFunctionNode() below?

---
In reply to: [68703551](https://github.com/Microsoft/ChakraCore/pull/1167#discussion_r68703551) [](ancestors = 68703551)

File: lib/Parser/Parse.cpp (hunk 2 above)
Comment (pleath) [inline lib/Parser/Parse.cpp]: Yes, that's what it means. The parser is now tracking func ID's and scope ID's separately for each name reference. A func ID != the declaration's func ID indicates a reference in a nested function. (And since a function's ID is < all its child ID's, we can assert that the reference ID is > the declaration's ID.)

File: lib/Parser/Parse.cpp (hunk 3 above)
Comment (pleath) [inline lib/Parser/Parse.cpp]: True. I put the comment in the right place.

File: lib/Parser/Parse.cpp (hunk 4 above)
Comment (pleath) [inline lib/Parser/Parse.cpp]: True. It was a little more painful to do that the way I originally structured the change, but it's easy now, so I'll do it.

File: lib/Parser/Parse.cpp (hunk 5 above)
Comment (pleath) [inline lib/Parser/Parse.cpp]: This is (supposed to be) the code from Parser::ParseFncDecl, moved in here because we now update the current function inside ParseFncDeclHelper.

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp (hunk 6 above)
Comment (pleath) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: It didn't matter before, because references did get bound properly, but we weren't retrieving the right scope. It does matter now, though.

File: lib/Runtime/ByteCode/ByteCodeGenerator.cpp (hunk 7 above)
Comment (pleath) [inline lib/Runtime/ByteCode/ByteCodeGenerator.cpp]: This is necessary for the change, yes. We need to do the merge earlier so we can propagate the has-local-in-closure property to the body scope before we process it.

File: lib/Runtime/ByteCode/ByteCodeEmitter.cpp [hunk 10]
```diff
@@ -3858,7 +3860,7 @@ void ByteCodeGenerator::StartEmitFunction(ParseNode *pnodeFnc)
                 // outer function's arguments directly.
                 sym = funcInfo->GetArgumentsSymbol();
                 Assert(sym);
-                if (sym->GetHasNonLocalReference() || autoCapturesAllScope.OldCapturesAll())
+                if (sym->NeedsSlotAlloc(funcInfo))
                 {
                     sym->EnsureScopeSlot(funcInfo);
```
Comment (aneeshdk) [inline lib/Runtime/ByteCode/ByteCodeEmitter.cpp]: EnsureScopeSlot() also does a NeedsSlotAlloc check.

**Comment by aneeshdk** [comment]
>                 top->SetFuncExprNameReference(true);

Shouldn't this be already true?

---
Refers to: lib/Runtime/ByteCode/ByteCodeGenerator.cpp:2455 in db13fe8. [](commit_id = db13fe85a2a4226e5774ba57f86e893841f00c78, deletion_comment = False)
